src/group_theory/monoid_localization.lean

```diff
@@ -406,8 +439,354 @@ lemma map_map {A} [comm_monoid A] {V} {g : Z →* A}
   map g hg (map f hf x) = map (g.comp f) (λ y : Y, hg ⟨f y, hf y⟩) x :=
 by rw ←map_comp_map hf hg; refl
 
-lemma map_ext (g : X →* Z) (hf : ∀ y : Y, f y ∈ W) (hg : ∀ y : Y, g y ∈ W)
+@[ext] lemma map_ext (g : X →* Z) (hf : ∀ y : Y, f y ∈ W) (hg : ∀ y : Y, g y ∈ W)
   (h : f = g) (x) : map f hf x = map g hg x :=
-induction_on x $ λ _, by {rw [map_mk, map_mk], congr; rw h; refl}
+induction_on x $ λ _,
+begin
+  rw [map_mk, map_mk],
+  congr;
+  rw h;
+  refl
+end
+
+open mul_equiv function
+
+/-- Given `comm_monoid`s `X, Z`, submonoids `Y ⊆ X, W ⊆ Z` and an isomorphism `h` of `X, Z` such
+    that `h(Y) = Z,` the isomorphism between `monoid_localization X Y` and
+    `monoid_localization Z W` induced by the map sending `x : X` to the equivalence class of
+    `(h x, 1)` in `monoid_localization Z W`. -/
+def mul_equiv_map (h : X ≃* Z) (H : h.to_monoid_hom.map Y = W) :
+  monoid_localization X Y ≃* monoid_localization Z W :=
+by refine { to_fun := map h.to_monoid_hom $ λ y : Y, show h y ∈ W, from
+    H ▸ set.mem_image_of_mem h y.2,
+  inv_fun := map h.symm.to_monoid_hom $ λ (w : W),
+    begin
+      obtain ⟨y, hym, hy⟩ : (w : Z) ∈ h.to_monoid_hom.map Y := H.symm ▸ w.2,
+      show h.symm w ∈ Y,
+      erw [hy.symm, symm_apply_apply],
+      exact hym,
+    end,
+  map_mul' := (map _ _).map_mul, ..};
+begin
+  intro x,
+  rw map_map,
+  conv_rhs {rw ←map_id x},
+  ext y,
+  simpa,
+end
+
+variables (f)
+
+/-- Given a `comm_monoid` `X` and a submonoid `Y ⊆ X`, the predicate characterizing `comm_monoid`s
+    isomorphic to the localization of `X` at `Y`. These are precisely the images of the
+    homomorphisms from the localization of `X` at `Y` induced by `comm_monoid` homomorphisms `f`
+    such that `f(Y)` consists of invertible elements, and for all `x, y : X`, `f(x) = f(y)` iff
```

Comment:
It is really quite unfortunate, I think, that we can't have this for additive monoids. It means that `nat → int` is not going to be an example of this predicate. But it is the most natural example of a monoid localization that exists in the real world. I wonder if we should rephrase the `is_unit` condition, to avoid talking about `units`, or alternatively, run `units` through the `to_additive` machine. @ChrisHughes24 @urkud any thoughts here?

Reply:
you can just use `additive` when applying the theory. We don't need to duplicate all of group theory on additive groups...

Reply:
I think that becomes really ugly for some think like `nat → int`, which is supposed to be the basic motivating example. It works, but that doesn't mean it's good.